src/kpro_connection.erl

```diff
@@ -384,9 +383,25 @@ maybe_reply({To, Ref}, Reply) ->
   _ = erlang:send(To, {Ref, Reply}),
   ok.
 
-loop(#state{} = State, Debug) ->
+loop(#state{draining = false} = State, Debug) ->
   Msg = receive Input -> Input end,
-  decode_msg(Msg, State, Debug).
+  decode_msg(Msg, State, Debug);
+
+loop(#state{draining = {_DrainState, DrainQueue}, requests = Requests} = State, Debug) ->
```

Comment:
maybe it can be achieved with less explicit states and less intrusive to the top-level loop call.
please consider this approach:

 `draining :: false | queue:queue()`.

1. `draining` is initialized to be `false`
2. When it's time to re-auth, init `draining = queue:from_list([reauth])`
3. Enqueue `{send, _} = Msg` when `draining` is not `false`
4. Handle all `draining` items when inflight request is empty

```
handle_msg(sasl_authenticate, State) ->
  State1 = State#state{backlog = queue:from_list([reauth])},
  %% maybe_handle_backlog should check if there are inflight requests
  %% if yes, do nothing, otherwise process the queue items
  State2 = maybe_handle_backlog(State1),
  ... continue looping with State2 ...
handle_msg({send, _} = Msg, #state{backlog = false}) ->
  ... handle as usual ...
handle_msg({send, _} = Msg, #state{backlog = Q} = State) ->
  NewState = State#state{backlog = queue:in(Q, Msg)},
  ... continue looping ...
```

then in this caluse: `handle_msg({_, Sock, Bin} ..) ->` 
it should call `maybe_handle_backlog` after `NewRequests = kpro_sent_reqs:del(Requests, CorrId),` but before looping.

Reply:
maybe also consider use `backlog` for the `draining` queue.
IMO, "draining" is for the inflight requests, but not the ones yet to be sent.

Reply:
Thanks for the feedback. 
I was initially hesitant to flush outside of the main loop, 
but in hindsight it is obvious that only the `{_, Socket, Bin}` handler can cause `requests` to become empty.

https://github.com/kafka4beam/kafka_protocol/compare/1ef51ba4330aeafd57bee4dd4869a15ba30c8a1b..f163a82d8c02cf47473fd3637d2eb05a4e738192